File: cpp-07/ex2/EliteCard.cpp
//
// Created by robot on 3/11/26.
//

#include "EliteCard.hpp"

std::map<std::string, std::variant<std::string, int>>
EliteCard::play(std::map<std::string, std::variant<std::string, int>> game_state)
{
    std::map<std::string, std::variant<std::string, int>> dict = {
            {"card_played", name},
            {"mana_used", cost},
            {"effect", std::string("Elite card entered the battlefield")}
    };
    return dict;
}

std::map<std::string, std::variant<std::string, int>> EliteCard::attack(const std::string &target)
{
    std::map<std::string, std::variant<std::string, int>> dict = {
            {"attacker", name},
            {"target", target},
            {"damage", attack_power},
            {"combat_type", std::string("melee")}
    };
    return dict;
}

std::map<std::string, std::variant<std::string, int>> EliteCard::defend(int incoming_damage)
{
    const int blocked = 3; // Fixed block for example
    int taken = std::max(0, incoming_damage - blocked);
    health -= taken;

    if (health > 0)
        health = 1;
    else
        health = 0;

    std::map<std::string, std::variant<std::string, int>> dict = {
            {"defender", name},
            {"damage_taken", taken},
            {"damage_blocked", blocked},
            {"still_alive", health}
    };
    return dict;
}

std::map<std::string, std::variant<std::string, int>> EliteCard::get_combat_stats() const
{
    std::map<std::string, std::variant<std::string, int>> dict = {
            {"attack", attack_power},
            {"health", health}
    };
    return dict;
}

std::map <std::string, std::variant<std::string, int>>
EliteCard::cast_spell(const std::string &spell_name, const std::vector <std::string> &targets)
{
    int mana_cost = 4;
    mana_pool -= mana_cost;
    std::map <std::string, std::variant<std::string, int>> dict = {
            {"caster", name},
            {"spell", spell_name},
            {"targets_count", (int)targets.size()},
            {"mana_used", mana_cost}
    };
    return dict;
}

std::map<std::string, std::variant<std::string, int>> EliteCard::channel_mana(int amount)
{
    mana_pool += amount;
    std::map<std::string, std::variant<std::string, int>> dict = {
            {"channeled", amount},
            {"total_mana", mana_pool}
    };
    return dict;
}

std::map<std::string, std::variant<std::string, int>> EliteCard::get_magic_stats() const
{
    std::map<std::string, std::variant<std::string, int>> dict = {
            {"mana_pool", mana_pool}
    };
    return dict;
}

File: python-07/ex2/EliteCard.py
from ex0.Card import Card
from .Combatable import Combatable
from .Magical import Magical


class EliteCard(Card, Combatable, Magical):

    def play(self, game_state: dict) -> dict:
        self.game_state = game_state

        return {
            'card_played': self.name,
            'mana_used': self.cost,
            'effect': 'Elite card summoned with combat and magic abilities',
            'card_type': 'elite'
        }

    def attack(self, target) -> dict:
        self.target = target
        return {
                "attacker": self.name,
                "target": self.target,
                "damage": self.cost,
                "combat_type": "melee"
                }

    def defend(self, incoming_damage: int) -> dict:
        self.incoming_damage = incoming_damage
        return {
                "defender": self.name,
                "damage_taken": incoming_damage - 2,
                "damage_blocked": incoming_damage - 1,
                "still_alive": True
                }

    def get_combat_stats(self) -> dict:
        pass

    def cast_spell(self, spell_name: str, targets: list) -> dict:
        self.spell_name = spell_name
        self.targets = targets
        return {
                "caster": self.name,
                "spell": self.spell_name,
                "targets":  self.targets,
                "mana_used": self.cost
                }

    def channel_mana(self, amount: int) -> dict:
        self.amount = amount
        return {"channeled": self.amount,
                "total_mana": self.amount + self.cost
                }

    def get_magic_stats(self) -> dict:
        return {
            'mana_pool': self._mana_pool,
            'current_mana': self._current_mana
        }
